File: src/teamlense_modules/agent/services/usage.service.ts
import { prisma } from "../../../shared/db/prisma";

export type ActivityCategory = "PRODUCTIVE" | "UNPRODUCTIVE" | "NEUTRAL";
export type ActivityTargetType = "APP" | "DOMAIN" | "URL";

type UsagePayload = {
  organizationId: string;
  userId: string;
  sessionId?: string;
  appName: string;
  windowTitle?: string;
  domain?: string;
  url?: string;
  durationSeconds: number;
  idleSeconds: number;
  isIdle: boolean;
  capturedAt: Date;
};

type RulePayload = {
  organizationId: string;
  targetType: ActivityTargetType;
  targetValue: string;
  category: ActivityCategory;
};

type SqlRow = Record<string, unknown>;

const productiveApps = [
  "visual studio code",
  "code.exe",
  "cursor",
  "intellij",
  "webstorm",
  "pycharm",
  "rider",
  "android studio",
  "xcode",
  "terminal",
  "windows terminal",
  "powershell",
  "git",
  "slack",
  "microsoft teams",
  "figma",
  "postman",
  "notion",
  "jira",
  "trello",
  "linear",
  "github",
  "gitlab",
  "bitbucket",
];

const productiveDomains = [
  "github.com",
  "gitlab.com",
  "bitbucket.org",
  "stackoverflow.com",
  "docs.microsoft.com",
  "developer.mozilla.org",
  "vercel.com",
  "linear.app",
  "jira.com",
  "atlassian.net",
  "notion.so",
  "figma.com",
  "slack.com",
  "teams.microsoft.com",
  "google.com",
];

const unproductiveDomains = [
  "youtube.com",
  "netflix.com",
  "primevideo.com",
  "hotstar.com",
  "instagram.com",
  "facebook.com",
  "x.com",
  "twitter.com",
  "reddit.com",
  "tiktok.com",
  "twitch.tv",
  "spotify.com",
];

const aiDomains = [
  "chatgpt.com",
  "openai.com",
  "claude.ai",
  "gemini.google.com",
  "copilot.microsoft.com",
  "perplexity.ai",
  "poe.com",
  "cursor.com",
];

const aiApps = ["chatgpt", "claude", "gemini", "copilot", "perplexity"];

const normalize = (value?: string | null): string => (value ?? "").trim().toLowerCase();

const asString = (value: unknown): string => {
  if (value instanceof Date) return value.toISOString();
  return typeof value === "string" ? value : "";
};

const asNumber = (value: unknown): number => {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : 0;
};

const toCategory = (value: unknown): ActivityCategory => {
  if (value === "PRODUCTIVE" || value === "UNPRODUCTIVE" || value === "NEUTRAL") return value;
  return "NEUTRAL";
};

const toTargetType = (value: unknown): ActivityTargetType => {
  if (value === "APP" || value === "DOMAIN" || value === "URL") return value;
  return "APP";
};

const domainMatches = (domain: string, candidates: string[]): boolean => {
  const clean = normalize(domain).replace(/^www\./, "");
  return candidates.some((candidate) => clean === candidate || clean.endsWith(`.${candidate}`));
};

export class UsageService {
  private static schemaReady = false;

  private static async ensureSchema(): Promise<void> {
    if (this.schemaReady || !prisma.$executeRawUnsafe) return;

    await prisma.$executeRawUnsafe(`
      DO $$
      BEGIN
        IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'ActivityCategory') THEN
          CREATE TYPE "ActivityCategory" AS ENUM ('PRODUCTIVE', 'UNPRODUCTIVE', 'NEUTRAL');
        END IF;
        IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'ActivityTargetType') THEN
          CREATE TYPE "ActivityTargetType" AS ENUM ('APP', 'DOMAIN', 'URL');
        END IF;
      END
      $$;
    `);

    await prisma.$executeRawUnsafe(`
      CREATE TABLE IF NOT EXISTS "activity_usage_logs" (
        "id" TEXT PRIMARY KEY,
        "organization_id" TEXT NOT NULL,
        "user_id" TEXT NOT NULL,
        "session_id" TEXT,
        "target_type" "ActivityTargetType" NOT NULL,
        "app_name" TEXT NOT NULL,
        "window_title" TEXT,
        "domain" TEXT,
        "url" TEXT,
        "category" "ActivityCategory" NOT NULL DEFAULT 'NEUTRAL',
        "duration_seconds" INTEGER NOT NULL DEFAULT 0,
        "idle_seconds" INTEGER NOT NULL DEFAULT 0,
        "is_idle" BOOLEAN NOT NULL DEFAULT false,
        "captured_at" TIMESTAMPTZ NOT NULL,
        "created_at" TIMESTAMPTZ NOT NULL DEFAULT NOW()
      )
    `);

    await prisma.$executeRawUnsafe(`
      CREATE TABLE IF NOT EXISTS "classification_rules" (
        "id" TEXT PRIMARY KEY,
        "organization_id" TEXT NOT NULL,
        "target_type" "ActivityTargetType" NOT NULL,
        "target_value" TEXT NOT NULL,
        "category" "ActivityCategory" NOT NULL,
        "created_at" TIMESTAMPTZ NOT NULL DEFAULT NOW(),
        "updated_at" TIMESTAMPTZ NOT NULL DEFAULT NOW()
      )
    `);

    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "active_application" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "window_title" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "domain" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "url" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "employee_name" TEXT`);
    await prisma.$executeRawUnsafe(`ALTER TABLE "screenshots" ADD COLUMN IF NOT EXISTS "project_name" TEXT`);

    await prisma.$executeRawUnsafe(
      `CREATE UNIQUE INDEX IF NOT EXISTS "classification_rules_organization_id_target_type_target_value_key"
       ON "classification_rules" ("organization_id", "target_type", "target_value")`,
    );
    await prisma.$executeRawUnsafe(
      `CREATE INDEX IF NOT EXISTS "activity_usage_logs_user_captured_idx"
       ON "activity_usage_logs" ("user_id", "captured_at")`,
    );
    await prisma.$executeRawUnsafe(
      `CREATE INDEX IF NOT EXISTS "activity_usage_logs_org_captured_idx"
       ON "activity_usage_logs" ("organization_id", "captured_at")`,
    );

    this.schemaReady = true;
  }

  static async classify(
    organizationId: string,
    targetType: ActivityTargetType,
    appName: string,
    domain?: string,
    url?: string,
  ): Promise<ActivityCategory> {
    await this.ensureSchema();

    const checks: Array<[ActivityTargetType, string | undefined]> = [
      ["URL", url],
      ["DOMAIN", domain],
      ["APP", appName],
    ];

    for (const [type, rawValue] of checks) {
      const targetValue = normalize(rawValue);
      if (!targetValue || !prisma.$queryRawUnsafe) continue;

      const rows = (await prisma.$queryRawUnsafe(
        `SELECT "category"
         FROM "classification_rules"
         WHERE "organization_id" = $1
           AND "target_type" = $2::"ActivityTargetType"
           AND "target_value" = $3
         LIMIT 1`,
        organizationId,
        type,
        targetValue,
      )) as SqlRow[];

      if (rows[0]) return toCategory(rows[0].category);
    }

    const app = normalize(appName);
    const cleanDomain = normalize(domain).replace(/^www\./, "");

    if (domainMatches(cleanDomain, aiDomains) || aiApps.some((candidate) => app.includes(candidate))) {
      return "NEUTRAL";
    }

    if (domainMatches(cleanDomain, unproductiveDomains)) return "UNPRODUCTIVE";
    if (domainMatches(cleanDomain, productiveDomains)) return "PRODUCTIVE";
    if (productiveApps.some((candidate) => app.includes(candidate))) return "PRODUCTIVE";

    return "NEUTRAL";
  }

  static async createUsageLog(payload: UsagePayload) {
    await this.ensureSchema();

    const domain = normalize(payload.domain) || null;
    const url = payload.url?.trim() || null;
    const targetType: ActivityTargetType = url ? "URL" : domain ? "DOMAIN" : "APP";
    const category = await this.classify(
      payload.organizationId,
      targetType,
      payload.appName,
      domain ?? undefined,
      url ?? undefined,
    );

    const durationSeconds = payload.isIdle ? 0 : Math.max(0, Math.round(payload.durationSeconds));
    const idleSeconds = Math.max(0, Math.round(payload.idleSeconds));

    if (prisma.$executeRawUnsafe) {
      await prisma.$executeRawUnsafe(
        `INSERT INTO "activity_usage_logs"
          ("id", "organization_id", "user_id", "session_id", "target_type", "app_name", "window_title",
           "domain", "url", "category", "duration_seconds", "idle_seconds", "is_idle", "captured_at", "created_at")
         VALUES ($1, $2, $3, $4, $5::"ActivityTargetType", $6, $7, $8, $9, $10::"ActivityCategory", $11, $12, $13, $14, NOW())`,
        crypto.randomUUID(),
        payload.organizationId,
        payload.userId,
        payload.sessionId ?? null,
        targetType,
        payload.appName,
        payload.windowTitle ?? null,
        domain,
        url,
        category,
        durationSeconds,
        idleSeconds,
        payload.isIdle,
        payload.capturedAt,
      );
    }

    return { category, targetType, durationSeconds };
  }

  static async upsertRule(payload: RulePayload) {
    await this.ensureSchema();
    const targetValue = normalize(payload.targetValue);

    const rows = (await prisma.$queryRawUnsafe(
      `INSERT INTO "classification_rules"
        ("id", "organization_id", "target_type", "target_value", "category", "created_at", "updated_at")
       VALUES ($1, $2, $3::"ActivityTargetType", $4, $5::"ActivityCategory", NOW(), NOW())
       ON CONFLICT ("organization_id", "target_type", "target_value")
       DO UPDATE SET "category" = EXCLUDED."category", "updated_at" = NOW()
       RETURNING "id", "target_type", "target_value", "category"`,
      crypto.randomUUID(),
      payload.organizationId,
      payload.targetType,
      targetValue,
      payload.category,
    )) as SqlRow[];

    return rows[0];
  }

  static async listRules(organizationId: string) {
    await this.ensureSchema();
    return prisma.$queryRawUnsafe(
      `SELECT "id", "target_type" AS "targetType", "target_value" AS "targetValue", "category"
       FROM "classification_rules"
       WHERE "organization_id" = $1
       ORDER BY "updated_at" DESC`,
      organizationId,
    ) as Promise<SqlRow[]>;
  }

  static async getUsageReport(params: {
    organizationId: string;
    userId?: string;
    start: Date;
    end: Date;
    groupBy?: "total" | "employee" | "team" | "location";
  }) {
    await this.ensureSchema();

    const userFilter = params.userId ? `AND l."user_id" = $4` : "";
    const values = params.userId
      ? [params.organizationId, params.start, params.end, params.userId]
      : [params.organizationId, params.start, params.end];

    const rows = (await prisma.$queryRawUnsafe(
      `SELECT
          COALESCE(NULLIF(l."domain", ''), l."app_name") AS "name",
          MAX(l."target_type"::text) AS "targetType",
          MAX(l."app_name") AS "appName",
          MAX(l."domain") AS "domain",
          MAX(l."category"::text) AS "category",
          SUM(l."duration_seconds")::int AS "durationSeconds",
          COUNT(*)::int AS "samples"
       FROM "activity_usage_logs" l
       WHERE l."organization_id" = $1
         AND l."captured_at" >= $2
         AND l."captured_at" <= $3
         ${userFilter}
       GROUP BY COALESCE(NULLIF(l."domain", ''), l."app_name")
       ORDER BY SUM(l."duration_seconds") DESC
       LIMIT 100`,
      ...values,
    )) as SqlRow[];

    const categoryRows = (await prisma.$queryRawUnsafe(
      `SELECT l."category"::text AS "category", SUM(l."duration_seconds")::int AS "durationSeconds"
       FROM "activity_usage_logs" l
       WHERE l."organization_id" = $1
         AND l."captured_at" >= $2
         AND l."captured_at" <= $3
         ${userFilter}
       GROUP BY l."category"
       ORDER BY SUM(l."duration_seconds") DESC`,
      ...values,
    )) as SqlRow[];

    return {
      items: rows.map((row) => ({
        name: asString(row.name),
        targetType: toTargetType(row.targetType),
        appName: asString(row.appName),
        domain: asString(row.domain),
        category: toCategory(row.category),
        durationSeconds: asNumber(row.durationSeconds),
        samples: asNumber(row.samples),
      })),
      categories: categoryRows.map((row) => ({
        category: toCategory(row.category),
        durationSeconds: asNumber(row.durationSeconds),
      })),
      groupBy: params.groupBy ?? "total",
    };
  }

}

File: backend-go/internal/services/usage_service.go
package services

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/teamlens/backend-go/internal/models"
)

type UsageService struct {
	pool *pgxpool.Pool
}

func NewUsageService(pool *pgxpool.Pool) *UsageService {
	return &UsageService{pool: pool}
}

var productiveApps = []string{
	"visual studio code", "code.exe", "cursor", "intellij", "webstorm",
	"pycharm", "rider", "android studio", "xcode", "terminal",
	"windows terminal", "powershell", "git", "slack", "microsoft teams",
	"figma", "postman", "notion", "jira", "trello", "linear",
	"github", "gitlab", "bitbucket", "excel", "winword", "powerpnt",
	"outlook", "zoom", "docker", "dbeaver", "tableplus", "datagrip",
	"sublime text", "notepad++", "obsidian",
}

var unproductiveApps = []string{
	"spotify", "netflix", "prime video", "instagram", "facebook",
	"tiktok", "steam", "epic games", "riot client", "valorant",
	"minecraft", "vlc", "media player",
}

var productiveDomains = []string{
	"github.com", "gitlab.com", "bitbucket.org", "stackoverflow.com",
	"docs.microsoft.com", "developer.mozilla.org", "vercel.com",
	"linear.app", "jira.com", "atlassian.net", "notion.so", "figma.com",
	"slack.com", "teams.microsoft.com", "google.com",
}

var unproductiveDomains = []string{
	"youtube.com", "netflix.com", "primevideo.com", "hotstar.com",
	"instagram.com", "facebook.com", "x.com", "twitter.com",
	"reddit.com", "tiktok.com", "twitch.tv", "spotify.com",
}

var aiDomains = []string{
	"chatgpt.com", "openai.com", "claude.ai", "gemini.google.com",
	"copilot.microsoft.com", "perplexity.ai", "poe.com", "cursor.com",
}

var aiApps = []string{"chatgpt", "claude", "gemini", "copilot", "perplexity"}

var appAliases = map[string]string{
	"code":          "Visual Studio Code",
	"code.exe":      "Visual Studio Code",
	"brave":         "Brave Browser",
	"brave.exe":     "Brave Browser",
	"chrome":        "Google Chrome",
	"chrome.exe":    "Google Chrome",
	"msedge":        "Microsoft Edge",
	"msedge.exe":    "Microsoft Edge",
	"firefox":       "Mozilla Firefox",
	"firefox.exe":   "Mozilla Firefox",
	"discord":       "Discord",
	"discord.exe":   "Discord",
}

var invalidDomainSuffixes = map[string]bool{
	"app": true, "css": true, "html": true, "js": true, "jsx": true,
	"json": true, "md": true, "py": true, "rs": true, "tsx": true,
	"ts": true, "txt": true, "vue": true, "xml": true,
}

func (s *UsageService) Classify(organizationID, targetType, appName, domain, url string) models.ActivityCategory {
	checks := []struct {
		targetType string
		value      string
	}{
		{"URL", url},
		{"DOMAIN", domain},
		{"APP", appName},
	}

	for _, check := range checks {
		if check.value == "" {
			continue
		}
		// Try to find in classification_rules
		var category string
		err := s.pool.QueryRow(context.Background(),
			`SELECT category::text
			 FROM classification_rules
			 WHERE organization_id = $1
			   AND target_type = $2::"ActivityTargetType"
			   AND target_value = $3
			 LIMIT 1`,
			organizationID, check.targetType, strings.ToLower(strings.TrimSpace(check.value)),
		).Scan(&category)
		if err == nil {
			return models.ActivityCategory(category)
		}
	}

	// Fall back to built-in classification
	appNormalized := strings.ToLower(strings.TrimSpace(appName))
	domainClean := cleanDomain(domain)

	if domainMatch(domainClean, aiDomains) || appMatchAny(appNormalized, aiApps) {
		return models.CatNeutral
	}
	if domainMatch(domainClean, unproductiveDomains) {
		return models.CatUnproductive
	}
	if domainMatch(domainClean, productiveDomains) {
		return models.CatProductive
	}
	if appMatchAny(appNormalized, unproductiveApps) {
		return models.CatUnproductive
	}
	if appMatchAny(appNormalized, productiveApps) {
		return models.CatProductive
	}

	return models.CatNeutral
}

func (s *UsageService) CreateUsageLog(ctx context.Context, payload *models.UsageLogPayload) (*models.UsageLogResult, error) {
	var domain string
	if payload.Domain != nil {
		domain = cleanDomain(*payload.Domain)
	}
	url := ""
	if payload.URL != nil {
		url = *payload.URL
	}

	targetType := "APP"
	if url != "" {
		targetType = "URL"
	} else if domain != "" {
		targetType = "DOMAIN"
	}

	category := s.Classify(payload.OrganizationID, targetType, payload.AppName, domain, url)

	durationSeconds := payload.DurationSeconds
	if durationSeconds < 0 {
		durationSeconds = 0
	}
	idleSeconds := payload.IdleSeconds
	if idleSeconds < 0 {
		idleSeconds = 0
	}

	_, err := s.pool.Exec(ctx,
		`INSERT INTO activity_usage_logs
		 (id, organization_id, user_id, session_id, target_type, app_name, window_title,
		  domain, url, category, duration_seconds, idle_seconds, is_idle, captured_at, created_at)
		 VALUES ($1, $2, $3, $4, $5::"ActivityTargetType", $6, $7, $8, $9, $10::"ActivityCategory", $11, $12, $13, $14, NOW())`,
		RandomToken(16),
		payload.OrganizationID,
		payload.UserID,
		payload.SessionID,
		targetType,
		payload.AppName,
		payload.WindowTitle,
		domain,
		url,
		category,
		durationSeconds,
		idleSeconds,
		payload.IsIdle,
		payload.CapturedAt,
	)
	if err != nil {
		return nil, fmt.Errorf("create usage log: %w", err)
	}

	return &models.UsageLogResult{
		Category:       category,
		TargetType:     targetType,
		DurationSeconds: durationSeconds,
	}, nil
}

func (s *UsageService) UpsertRule(ctx context.Context, organizationID string, input *models.UpsertRuleInput) (*models.ClassificationRule, error) {
	targetValue := strings.ToLower(strings.TrimSpace(input.TargetValue))

	var rule models.ClassificationRule
	err := s.pool.QueryRow(ctx,
		`INSERT INTO classification_rules (id, organization_id, target_type, target_value, category, created_at, updated_at)
		 VALUES ($1, $2, $3::"ActivityTargetType", $4, $5::"ActivityCategory", NOW(), NOW())
		 ON CONFLICT (organization_id, target_type, target_value)
		 DO UPDATE SET category = EXCLUDED.category, updated_at = NOW()
		 RETURNING id, target_type::text, target_value, category::text`,
		RandomToken(16), organizationID, input.TargetType, targetValue, input.Category,
	).Scan(&rule.ID, &rule.TargetType, &rule.TargetValue, &rule.Category)
	if err != nil {
		return nil, fmt.Errorf("upsert rule: %w", err)
	}
	return &rule, nil
}

func (s *UsageService) ListRules(ctx context.Context, organizationID string) ([]models.ClassificationRule, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, target_type::text, target_value, category::text
		 FROM classification_rules
		 WHERE organization_id = $1
		 ORDER BY updated_at DESC`,
		organizationID,
	)
	if err != nil {
		return nil, fmt.Errorf("list rules: %w", err)
	}
	defer rows.Close()

	var rules []models.ClassificationRule
	for rows.Next() {
		var r models.ClassificationRule
		if err := rows.Scan(&r.ID, &r.TargetType, &r.TargetValue, &r.Category); err != nil {
			return nil, fmt.Errorf("scan rule: %w", err)
		}
		rules = append(rules, r)
	}
	return rules, nil
}

func (s *UsageService) DeleteRule(ctx context.Context, organizationID, ruleID string) error {
	_, err := s.pool.Exec(ctx,
		`DELETE FROM classification_rules WHERE id = $1 AND organization_id = $2`,
		ruleID, organizationID,
	)
	return err
}

func cleanDomain(value string) string {
	domain := strings.ToLower(strings.TrimSpace(value))
	domain = strings.TrimPrefix(domain, "https://")
	domain = strings.TrimPrefix(domain, "http://")
	domain = strings.TrimPrefix(domain, "www.")
	if idx := strings.IndexAny(domain, "/?#"); idx >= 0 {
		domain = domain[:idx]
	}
	if !isValidDomain(domain) {
		return ""
	}
	return domain
}

func isValidDomain(value string) bool {
	if value == "" {
		return false
	}
	parts := strings.Split(value, ".")
	if len(parts) < 2 {
		return false
	}
	suffix := parts[len(parts)-1]
	if invalidDomainSuffixes[suffix] {
		return false
	}
	if strings.Contains(value, "..") {
		return false
	}
	for _, p := range parts {
		if p == "" {
			return false
		}
	}
	return true
}

func domainMatch(domain string, candidates []string) bool {
	if domain == "" {
		return false
	}
	for _, c := range candidates {
		if domain == c || strings.HasSuffix(domain, "."+c) {
			return true
		}
	}
	return false
}

func appMatchAny(app string, candidates []string) bool {
	app = strings.ToLower(app)
	for _, c := range candidates {
		if strings.Contains(app, c) {
			return true
		}
	}
	return false
}

func (s *UsageService) GetUsageReport(ctx context.Context, params struct {
	OrganizationID string
	UserID         *string
	TeamID         *string
	Start          time.Time
	End            time.Time
	GroupBy        string
}) (*models.UsageReport, error) {
	// Build a simplified usage report
	rows, err := s.pool.Query(ctx,
		`SELECT COALESCE(NULLIF(aul.domain, ''), aul.app_name) AS name,
		        COALESCE(MAX(aul.target_type::text), 'APP') AS target_type,
		        COALESCE(MAX(aul.app_name), '') AS app_name,
		        COALESCE(MAX(aul.domain), '') AS domain,
		        MAX(aul.category::text) AS category,
		        SUM(aul.duration_seconds)::int AS duration_seconds,
		        COUNT(*)::int AS samples
		 FROM activity_usage_logs aul
		 WHERE aul.organization_id = $1
		   AND aul.captured_at >= $2
		   AND aul.captured_at <= $3
		   AND ($4::text IS NULL OR aul.user_id = $4::text)
		   AND ($5::text IS NULL OR EXISTS(
		       SELECT 1 FROM team_memberships tm WHERE tm.user_id = aul.user_id AND tm.team_id = $5::text
		   ))
		 GROUP BY COALESCE(NULLIF(aul.domain, ''), aul.app_name)
		 ORDER BY SUM(aul.duration_seconds) DESC
		 LIMIT 100`,
		params.OrganizationID, params.Start, params.End, params.UserID, params.TeamID,
	)
	if err != nil {
		return nil, fmt.Errorf("query usage: %w", err)
	}
	defer rows.Close()

	var items []models.UsageReportItem
	for rows.Next() {
		var item models.UsageReportItem
		if err := rows.Scan(&item.Name, &item.TargetType, &item.AppName, &item.Domain,
			&item.Category, &item.DurationSeconds, &item.Samples); err != nil {
			return nil, fmt.Errorf("scan usage item: %w", err)
		}
		items = append(items, item)
	}

	if items == nil {
		items = []models.UsageReportItem{}
	}

	categoryRows, err := s.pool.Query(ctx,
		`SELECT aul.category::text AS name,
		        MAX(aul.category::text) AS category,
		        SUM(aul.duration_seconds)::int AS duration_seconds
		 FROM activity_usage_logs aul
		 WHERE aul.organization_id = $1
		   AND aul.captured_at >= $2
		   AND aul.captured_at <= $3
		   AND ($4::text IS NULL OR aul.user_id = $4::text)
		   AND ($5::text IS NULL OR EXISTS(
		       SELECT 1 FROM team_memberships tm WHERE tm.user_id = aul.user_id AND tm.team_id = $5::text
		   ))
		 GROUP BY aul.category
		 ORDER BY SUM(aul.duration_seconds) DESC`,
		params.OrganizationID, params.Start, params.End, params.UserID, params.TeamID,
	)
	if err != nil {
		return nil, fmt.Errorf("query usage categories: %w", err)
	}
	defer categoryRows.Close()

	var categories []models.UsageCategoryBreakdown
	for categoryRows.Next() {
		var item models.UsageCategoryBreakdown
		if err := categoryRows.Scan(&item.Name, &item.Category, &item.DurationSeconds); err != nil {
			return nil, fmt.Errorf("scan usage category: %w", err)
		}
		categories = append(categories, item)
	}
	if categories == nil {
		categories = []models.UsageCategoryBreakdown{}
	}

	breakdownRows, err := s.pool.Query(ctx,
		`SELECT COALESCE(NULLIF(aul.domain, ''), aul.app_name) AS name,
		        COALESCE(u.full_name, 'Unknown employee') AS employee_name,
		        COALESCE(t.name, 'No team') AS team_name,
		        COALESCE(NULLIF(ws.location_type, ''), 'Unknown') AS location_name,
		        SUM(aul.duration_seconds)::int AS duration_seconds,
		        COUNT(*)::int AS samples
		 FROM activity_usage_logs aul
		 LEFT JOIN users u ON u.id = aul.user_id
		 LEFT JOIN work_sessions ws ON ws.id = aul.session_id
		 LEFT JOIN team_memberships tm ON tm.user_id = aul.user_id
		 LEFT JOIN teams t ON t.id = tm.team_id
		 WHERE aul.organization_id = $1
		   AND aul.captured_at >= $2
		   AND aul.captured_at <= $3
		   AND ($4::text IS NULL OR aul.user_id = $4::text)
		   AND ($5::text IS NULL OR tm.team_id = $5::text)
		 GROUP BY COALESCE(NULLIF(aul.domain, ''), aul.app_name), u.full_name, t.name, ws.location_type
		 ORDER BY SUM(aul.duration_seconds) DESC
		 LIMIT 500`,
		params.OrganizationID, params.Start, params.End, params.UserID, params.TeamID,
	)
	if err != nil {
		return nil, fmt.Errorf("query usage breakdowns: %w", err)
	}
	defer breakdownRows.Close()

	var breakdowns []models.UsageBreakdownItem
	for breakdownRows.Next() {
		var item models.UsageBreakdownItem
		if err := breakdownRows.Scan(&item.Name, &item.EmployeeName, &item.TeamName, &item.LocationName, &item.DurationSeconds, &item.Samples); err != nil {
			return nil, fmt.Errorf("scan usage breakdown: %w", err)
		}
		breakdowns = append(breakdowns, item)
	}
	if breakdowns == nil {
		breakdowns = []models.UsageBreakdownItem{}
	}

	return &models.UsageReport{
		Items:      items,
		Categories: categories,
		Breakdowns: breakdowns,
		GroupBy:    params.GroupBy,
	}, nil
}
